File: BOJ/31_트리에서의동적계획법/2533_사회망서비스.cpp
#include <stdio.h>
#include <vector>
#include <algorithm>

#define MAXN 1000000

using namespace std;

int N,x,y,visitied[MAXN+1],dp[MAXN+1][2];
vector<int> adjs[MAXN+1];

void dfs(int cur) {
    visitied[cur] = 1;

    dp[cur][0] = 0;
    dp[cur][1] = 1;

    for(int i=0; i<adjs[cur].size(); i++) {
        int next = adjs[cur][i];
        if(visitied[next] != 0) continue;
        dfs(next);

        dp[cur][0] += dp[next][1];
        dp[cur][1] += min(dp[next][0],dp[next][1]);
    }
}

int main() {

    scanf("%d",&N);
    for(int i=0; i<N; i++) {
        scanf("%d %d",&x,&y);
        adjs[x].push_back(y);
        adjs[y].push_back(x);
    }
    dfs(1);

    printf("%d\n",min(dp[1][0],dp[1][1]));
    
    return 0;
}

File: BOJ/31_트리에서의동적계획법/2533_사회망서비스.py
import sys

read = sys.stdin.readline
sys.setrecursionlimit(int(1e9))
def dfs(cur) :
    visited[cur] = 1

    dp[cur][0] = 0
    dp[cur][1] = 1

    for nextNode in adjs[cur] :
        if visited[nextNode] != 0 : continue
        dfs(nextNode)

        dp[cur][0] += dp[nextNode][1]
        dp[cur][1] += min(dp[nextNode][0],dp[nextNode][1])

N = int(read())
adjs = [[] for _ in range(N+1)]
visited = [0 for _ in range(N+1)]
dp = [[0,0] for _ in range(N+1)]
for _ in range(N-1) :
    x,y = map(int,read().split())
    adjs[x].append(y)
    adjs[y].append(x)

dfs(1)

print(min(dp[1][0],dp[1][1]))